LG-356 Add help text to the account creation screen for SAM

stevegsa opened the pull request:
**Why**: To give SAM users clear instructions upon account creation

**How**: Conditionally show the sp alert partial and change it's contents for SAM

Hi! Before submitting your PR for review, and/or before merging it, please
go through the following checklist:

- [x] For DB changes, check for missing indexes, check to see if the changes
affect other apps (such as the dashboard), make sure the DB columns in the
various environments are properly populated, coordinate with devops, plan
migrations in separate steps.

- [x] For route changes, make sure GET requests don't change state or result in
destructive behavior. GET requests should only result in information being
read, not written.

- [x] For encryption changes, make sure it is compatible with data that was
encrypted with the old code.

- [x] For secrets changes, [make sure to update the S3 secrets bucket](https://github.com/18F/identity-private/wiki/Secrets-S3-buckets) with the 
new configs in **all** environments. 

- [x] Do not disable Rubocop or Reek offenses unless you are absolutely sure
they are false positives. If you're not sure how to fix the offense, please
ask a teammate.

- [x] When reading data, write tests for nil values, empty strings,
and invalid formats.

- [x] When calling `redirect_to` in a controller, use `_url`, not `_path`.

- [x] When adding user data to the session, use the `user_session` helper
instead of the `session` helper so the data does not persist beyond the user's
session.

- [x] When adding a new controller that requires the user to be fully
authenticated, make sure to add `before_action :confirm_two_factor_authenticated`.

monfresh commented:
I thought we already had this enabled on the account creation screen for all SPs, but I guess not. Is there a reason why we wouldn't want this on the account creation screen for all SPs?

stevegsa commented:
I spoke to Vishal.  The reason we are only doing it for SAM is because the wording is specific to them vs say CBP/TTP (which doesn't need to use the same email address).  So rather than require usajobs and ttp, etc. to provide new wording for this screen we are hardcoding it to the one that needs this feature now.

monfresh commented:
Isn't that why we have SP-specific entries in the YAML file, which allows us to customize the message for each agency? The wording is already correct for the other agencies. Basically, what I'm saying is we already have text in place that shows up for all agencies on the `sign_up/start` page, and we want the same message on the "enter your email" page.

vishaliyer-18f commented:
The intent and content of the messaging in the create account state is different from the landing state & sign in. There is some overlap, but they are essentially different.

In TTP for e.g, we have "Your old GOES userID and password won’t work. Please create a login.gov account below." for landing & sign in. This doesn't really apply for create account because they are already doing that. So, there's no need for any additional text for CBP in the create account state.

monfresh commented:
Yes, but for USAJOBS, it is relevant since we want to make sure they use the same email address. From an engineering perspective, instead of adding conditionals to the view, I would add the appropriate entries in the YAML file and leave them blank for those agencies that don't need any messaging on the account creation screen.

stevegsa commented:
I spoke to Joel about refactoring this code in general before I changed it.  Specifically we want to take out all references in the code to SPs (ie SP_ALERTS) and have it data driven.  For now I just added to the mess rather than refactoring the whole thing.

stevegsa commented:
If you would prefer I could make that one conditional dependent on if the text is populated and not have the hardcoded sp.  However this code still needs to be refactored.

stevegsa commented:
The "if current_page"... code also needs to go.

vishaliyer-18f commented:
@monfresh, you are correct for USAJOBS. The desired goal is to enable- 
1. Different copy from landing/ sign in states 
2. The text being shown only for some SP's on the create account page.

stevegsa commented:
I went ahead and added the code to conditionally show the help text on the account creation screen if they supply it.

monfresh commented on app/views/shared/_sp_alert.html.slim:
```diff
@@ -1,10 +1,13 @@
-  - sp_alert_name = decorated_session.sp_alert_name
+- sp_alert_name = decorated_session.sp_alert_name
+- has_sign_up_text = I18n.exists?("service_providers.#{sp_alert_name}.create_account_page.body")
+- if decorated_session.sp_alert? && !(current_page?(sign_up_email_path) && !has_sign_up_text)
```
Why not create an empty `create_account_page.body` entry for those SPs that don't need one? This will eliminate these conditionals which are hard to read IMO. It is generally considered a bad practice to have conditionals in views. We prefer the "tell don't ask" principle, such as in our heavy use of presenters, which could be another option here.

stevegsa replied:
That would imply creating a blank entry for every sp in the yml file as well as every future sp and translation when on average it will not be supplied.  I agree that it's bad style to put these conditionals in the view but the conditionals were there before I got here. The correct way to do this would be to pull out all the sp specific code from the views and session decorator (we already have the issuer which is unique) and drive it with some data configuration that feeds the decorated session which feeds the view. You are proposing to use the locale files as a mechanism to configure what gets shown.  I can do it but I'd prefer 2 lines of code over dozens of unused lines in the locale files.

monfresh replied:
Fair enough.

stevegsa replied:
I'm not totally against it.  At least it makes it more data bound.  I think the specs would need to be updated because they are testing for an alert box.  I'm not sure if the alert box is invisible or not when there is no text.

monfresh commented on config/locales/service_providers/fr.yml:
```diff
@@ -28,3 +28,7 @@ fr:
+      create_account_page:
+        body: Si vous avez déjà un profil SAM, veuillez utiliser l'adresse e-mail
+          principale ou secondaire que vous avez utilisée pour SAM pour créer votre
+          nouveau compte login.gov.
```
Why do the French and Spanish translations refer to "primary or secondary" email address, whereas the English test does not?

stevegsa replied:
I was supposed to use the text from account_page.body for create_account_page.body but it appears the account_page.body for spanish and french are currently wrong in production.  I need new translations for both.

monfresh commented on config/locales/service_providers/fr.yml:
```diff
@@ -18,13 +20,18 @@ fr:
-          nouveau compte login.gov.
+        body: Votre ancien nom d'utilisateur et mot de passe SAM ne fonctionnera pas.
+          S'il vous plaît créer un login.gov compte en utilisant la même adresse e-mail
+          que vous utilisez pour SAM.
```
Please use this instead: `Votre ancien nom d'utilisateur et mot de passe SAM ne marchera pas.
          Veuillez créer un nouveau compte login.gov avec la même adresse e-mail que vous avez utilisée pour SAM.`

monfresh commented on config/locales/service_providers/fr.yml:
```diff
@@ -18,13 +20,18 @@ fr:
       create_account_link: créer votre nouveau compte login.gov
+      create_account_page:
+        body: S'il vous plaît créer un login.gov compte en utilisant la même adresse
+          e-mail que vous utilisez pour SAM.
```
Please use this text: `Veuillez créer un compte login.gov avec la même adresse e-mail que vous avez utilisée pour SAM.`
